### ISSUE TITLE
TST: --open-files did not catch ResourceWarning

### ISSUE THREAD

**Issue by pllim** [astropy/astropy#7913]
Is this is not really an issue, feel free to close, as I am not familiar with the internals of open files check. Example log: https://travis-ci.org/astropy/astropy/jobs/442639768

It seems weird to me that `--open-files` option still reported passing despite the presence of following warning:

```
sys:1: ResourceWarning: unclosed file <_io.BufferedWriter name=13>
```

cc @drdavella

**Comment by astrofrog** [comment]
Maybe this issue should be at https://github.com/astropy/pytest-openfiles?

**Comment by pllim** [comment]
After some digging... could this be #6942 ?

**Comment by pllim** [comment]
Moved to astropy/pytest-openfiles#14

**Comment by drdavella** [comment]
I have some reason to believe this warning is being caused by this particular test, although I'm not sure why, and it doesn't really help much yet. (Is anyone surprised that it comes from `io.fits` 😝?)

https://github.com/astropy/astropy/blob/66c79673752a5771f30de135a2c8b861b3892995/astropy/io/fits/tests/test_fitsheader.py#L16-L21

It's also entirely possible that this is just a red herring.

**Comment by drdavella** [comment]
So the `ResourceWarning` is definitely being caused by an oversight in `astropy/io/fits/scripts/fitsheader.py`. I can make it go away by adding a `close()` in the right place.

The real question still is why isn't getting picked up by `pytest-openfiles`?

**Comment by drdavella** [comment]
I think this is not a problem with `pytest-openfiles`, but there's something really strange going on, and I'm running out of ideas. I think this distills the problem nicely:

```python
#!/usr/bin/env python -W error::ResourceWarning

import psutil
from astropy.io.fits.scripts import fitsheader

p = psutil.Process()

print('before', p.open_files())
fitsheader.main(['astropy/io/fits/tests/data/arange.fits'])
print('after', p.open_files())
```

And here's the output:

```
before []
# HDU 0 in astropy/io/fits/tests/data/arange.fits:
SIMPLE  =                    T / conforms to FITS standard
BITPIX  =                   32 / array data type
NAXIS   =                    3 / number of array dimensions
NAXIS1  =                   11
NAXIS2  =                   10
NAXIS3  =                    7
EXTEND  =                    T
Exception ignored in: <_io.FileIO name='astropy/io/fits/tests/data/arange.fits' mode='rb' closefd=True>
ResourceWarning: unclosed file <_io.FileIO name='astropy/io/fits/tests/data/arange.fits' mode='rb' closefd=True>
after []
```

The key takeaway is that even though we get a resource warning for the file being open at the end of the process, it does *not* show up in the list of open files for this process using `psutil`. I guess it's possible that the file handle is owned by another process (???) and so looking at the open files for this process is not productive. However, assuming this is actually the problem (and I really don't know), I'm not sure we can handle this in a systematic way in `pytest-openfiles`.

It's fairly easy to fix this particular issue in `fitsheader`. But I'm not sure how we could have detected it automatically.

**Comment by pllim** [comment]
Thanks for the detail investigation! I have re-opened this issue.

**Comment by pllim** [comment]
While looking into #7928, I might have found one of the cause:
```
astropy/coordinates/tests/test_solar_system.py::TestPositionsGeocentric::()::test_de432s_planet[mercury]
  .../astropy/coordinates/solar_system.py:364: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    get_velocity=False)

astropy/coordinates/tests/test_solar_system.py::TestPositionsGeocentric::()::test_de432s_planet[jupiter]
  .../astropy/coordinates/solar_system.py:364: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    get_velocity=False)

astropy/coordinates/tests/test_solar_system.py::TestPositionsGeocentric::()::test_de432s_planet[sun]
  .../astropy/coordinates/solar_system.py:364: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    get_velocity=False)

astropy/coordinates/tests/test_solar_system.py::TestPositionsGeocentric::()::test_de432s_moon
  .../astropy/coordinates/solar_system.py:364: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    get_velocity=False)

astropy/coordinates/tests/test_solar_system.py::TestPositionKittPeak::()::test_de432s_planet[mercury]
  .../astropy/coordinates/solar_system.py:364: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    get_velocity=False)

astropy/coordinates/tests/test_solar_system.py::TestPositionKittPeak::()::test_de432s_planet[jupiter]
  .../astropy/coordinates/solar_system.py:364: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    get_velocity=False)

astropy/coordinates/tests/test_solar_system.py::TestPositionKittPeak::()::test_de432s_moon
  .../astropy/coordinates/solar_system.py:364: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    get_velocity=False)

astropy/coordinates/tests/test_solar_system.py::TestPositionKittPeak::()::test_custom_kernel_spec_body[mercury]
  .../astropy/coordinates/solar_system.py:364: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    get_velocity=False)

astropy/coordinates/tests/test_solar_system.py::TestPositionKittPeak::()::test_custom_kernel_spec_body[jupiter]
  .../astropy/coordinates/solar_system.py:364: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    get_velocity=False)

astropy/coordinates/tests/test_solar_system.py::test_get_sun_consistency[time0]
  .../astropy/coordinates/solar_system.py:364: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    get_velocity=False)

astropy/coordinates/tests/test_solar_system.py::test_get_sun_consistency[time1]
  .../astropy/coordinates/solar_system.py:364: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    get_velocity=False)

astropy/coordinates/tests/test_solar_system.py::test_get_sun_consistency[time2]
  .../astropy/coordinates/solar_system.py:364: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    get_velocity=False)

astropy/coordinates/tests/test_solar_system.py::test_barycentric_velocity_consistency[mercury-pos_tol0-vel_tol0]
  .../astropy/coordinates/solar_system.py:330: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    return _get_body_barycentric_posvel(body, time, ephemeris)

astropy/coordinates/tests/test_solar_system.py::test_barycentric_velocity_consistency[jupiter-pos_tol1-vel_tol1]
  .../astropy/coordinates/solar_system.py:330: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    return _get_body_barycentric_posvel(body, time, ephemeris)

astropy/coordinates/tests/test_solar_system.py::test_barycentric_velocity_consistency[earth-pos_tol2-vel_tol2]
  .../astropy/coordinates/solar_system.py:330: ResourceWarning: unclosed file <_io.BufferedReader name='...cache/astropy/download/py3/d66c2153188958472fec67016b4b4e7a'>
    return _get_body_barycentric_posvel(body, time, ephemeris)
```

@eteq or @adrn , any suggestion on how to fix?

**Comment by pllim** [comment]
Wait, maybe this particular one reported originally is from `io.fits`...

**Comment by drdavella** [comment]
IIRC, I think the problem is that many of them come from other libraries. If that's the case, we would want to somehow filter to get only the ones that come from `astropy`, but I'm not sure there's a straightforward way to do that in Python.

That's one of the reasons I haven't really made any progress on implementing this functionality in `pytest-openfiles`, although it would still be really useful.

**Comment by saimn** [comment]
Actually there is a way to show a traceback for ResourceWarning, with tracemalloc: `python3 -Wd -X tracemalloc=5`. From https://pythondev.readthedocs.io/debug_tools.html#resourcewarning, and thanks to @vstinner !

**Comment by jdavies-st** [comment]
Re https://github.com/astropy/astropy/issues/7913#issuecomment-431538395 above, when you get a `ResourceWarning` for an open file from python, it's because the garbage collector detected the unclosed file, closed it and printed that warning, so it is not surprising that `p.open_files()` is empty before and after.

**Comment by bsipocz** [comment]
addressed by #8439
